src/type-layout.md

```diff
@@ -217,31 +248,65 @@ without a `repr` attribute. Using this representation explicitly through a
 
- 1. The fields are properly aligned.
- 2. The fields do not overlap.
- 3. The alignment of the type is at least the maximum alignment of its fields.
+#### Discriminant Ellision
 
-r[layout.repr.rust.alignment]
-Formally, the first guarantee means that the offset of any field is divisible by
-that field's alignment.
+r[layout.repr.rust.option]
 
-r[layout.repr.rust.field-storage]
-The second guarantee means that the fields can be
-ordered such that the offset plus the size of any field is less than or equal to
-the offset of the next field in the ordering. The ordering does not have to be
-the same as the order in which the fields are specified in the declaration of
-the type.
+r[layout.repr.rust.option.intro]
+Certain `repr(Rust)` enums are specially layed out when used with certain types. This layout process is known as discriminant elision.
 
-Be aware that the second guarantee does not imply that the fields have distinct
-addresses: zero-sized types may have the same address as other fields in the
-same struct.
+r[layout.repr.rust.option.elligible]
+An `enum` type is a *discriminant elision eligible* enum if:
+* It has exactly two variants,
+* One variant has exactly one field, known as the *elision candidate field*, and
+* The other variant has no fields that have size greater than 0 or alignment greater than 1, known as the *elided variant*.
 
-r[layout.repr.rust.unspecified]
-There are no other guarantees of data layout made by this representation.
+> [!NOTE]
+> The determination of whether an `enum` type is *discriminant elision eligible* applies after mono-morphization.
+> In particular, a type like [`core::result::Result<T,E>`] is *discriminant elision eligible* if either `T` or `E` is instantiated with a 1-ZST type, and [`core::option::Option<T>`] is *discriminant elision eligible* always (due to its `None` variant).
+
+r[layout.repr.rust.option.candidates]
+The following types are *elision candidate types*:
+* [`&U`][type.pointer.reference.shared], if `U` is a `Sized` type
+* [`&mut U`][type.pointer.reference.mut], if `U` is a `Sized` type
+* [`core::ptr::NonNull<U>`], if `U` is a `Sized` type
+* [`alloc::boxed::Box<U>`], if `U` is a `Sized` type
+* [`core::num::NonZero<T>`] if `T` is an integer type
+* A [function pointer][type.fn-pointer] type
+* A `struct`, other than [`core::cell::UnsafeCell<T>`], defined using [`repr(transparent)`][layout.repr.transparent], which has a field that is an *elision candidate*, and all other fields have size 0 and alignment 1.
+
+r[layout.repr.rust.option.elision]
+If the *elision candidate field* of a *discriminant elision eligible* `enum` has an *elision candidate type*, then the `enum` has the same layout as that field, except that the value consisting of all `0` bytes represents the *elided variant* of the `enum`.
```

Comment:
This part is strictly incorrect. It will have the _associated value_, determined by the "inner type" `T`, as T-lang has already FCPed a type that is all-`0xFF` bytes.